File: apps/api/app/expenses/application/update_expense.py
"""Use case: edit an existing expense (partial update)."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date
from typing import Any, TypeVar

from app.expenses.domain.entities import Expense
from app.expenses.domain.errors import ExpenseNotFoundError, UnknownCategoryError
from app.expenses.ports.category_checker import CategoryChecker
from app.expenses.ports.repository import ExpenseRepository
from app.shared.domain.money import Money

_T = TypeVar("_T")


@dataclass(frozen=True, slots=True)
class UpdateExpenseCommand:
    expense_id: int
    fields: frozenset[str]  # which attributes the caller intends to change
    amount_cents: int | None = None
    currency: str | None = None
    category_id: int | None = None
    occurred_on: date | None = None
    note: str | None = None


class UpdateExpense:
    def __init__(self, repository: ExpenseRepository, categories: CategoryChecker) -> None:
        self._repository = repository
        self._categories = categories

    def __call__(self, command: UpdateExpenseCommand) -> Expense:
        existing = self._repository.get(command.expense_id)
        if existing is None:
            raise ExpenseNotFoundError(command.expense_id)

        amount_cents = self._pick(command, "amount_cents", existing.money.amount_cents)
        currency = self._pick(command, "currency", existing.money.currency)
        category_id = self._pick(command, "category_id", existing.category_id)
        occurred_on = self._pick(command, "occurred_on", existing.occurred_on)
        note = self._pick(command, "note", existing.note)

        if "category_id" in command.fields and not self._categories.exists(category_id):
            raise UnknownCategoryError(category_id)

        updated = Expense(
            id=existing.id,
            money=Money(amount_cents, currency),
            category_id=category_id,
            occurred_on=occurred_on,
            note=note,
        )
        return self._repository.update(updated)

    @staticmethod
    def _pick(command: UpdateExpenseCommand, name: str, current: _T) -> _T:
        value: Any = getattr(command, name)
        return value if name in command.fields else current

File: apps/api/internal/expenses/application/update_expense.go
package application

import (
	"context"

	"caudal-api/internal/expenses/domain"
	"caudal-api/internal/expenses/ports"
	shared "caudal-api/internal/shared/domain"
)

type UpdateExpense struct {
	repository ports.Repository
	checker    ports.CategoryChecker
}

func NewUpdateExpense(repository ports.Repository, checker ports.CategoryChecker) UpdateExpense {
	return UpdateExpense{repository: repository, checker: checker}
}

func (useCase UpdateExpense) Execute(ctx context.Context, command UpdateExpenseCommand) (domain.Expense, error) {
	expense, err := useCase.repository.Get(ctx, command.ID)
	if err != nil {
		return domain.Expense{}, err
	}
	if err := useCase.applyChanges(ctx, &expense, command); err != nil {
		return domain.Expense{}, err
	}
	return useCase.repository.Update(ctx, expense)
}

func (useCase UpdateExpense) applyChanges(ctx context.Context, expense *domain.Expense, command UpdateExpenseCommand) error {
	if command.CategoryID != nil {
		exists, err := useCase.checker.ExpenseCategoryExists(ctx, *command.CategoryID)
		if err != nil || !exists {
			if err != nil {
				return err
			}
			return domain.ErrUnknownCategory
		}
		expense.CategoryID = *command.CategoryID
	}
	if command.AmountCents != nil || command.Currency != nil {
		amount, currency := expense.Money.AmountCents, expense.Money.Currency
		if command.AmountCents != nil {
			amount = *command.AmountCents
		}
		if command.Currency != nil {
			currency = *command.Currency
		}
		money, err := shared.NewMoney(amount, currency)
		if err != nil {
			return err
		}
		expense.Money = money
	}
	if command.OccurredOn != nil {
		expense.OccurredOn = *command.OccurredOn
	}
	if command.Note != nil {
		expense.Note = *command.Note
	}
	if command.RecurrenceID != nil {
		expense.RecurrenceID = *command.RecurrenceID
	}
	return nil
}
